### ISSUE TITLE
Time Elapsing Breaks Wallet Identities

### ISSUE THREAD

**Issue by ScottMorris** [hyperledger/composer#1731]
While working with the composer rest api and developing the business network it seems that after an unknown period of time elapses the identities assigned to a wallet no longer work and new ones have be issues and re-associated to the participants and the authenticated user's wallet.

## Context

I'm assuming that the identities should persist through time.

## Expected Behaviour
Identity relationships should persist across time.

## Actual Behavior
When a network update is performed and an endpoint is called, eg. _ping_ an error is returned.

The same thing happens in both the CLI and the REST API.
```
scott@ubuntu-hyperledger2:~$ composer network ping -p hlfv1 -n mynetwork-i Amy -s CqqCUerzNHXe

Error: Error trying to ping. Error: Error trying to query chaincode. Error: chaincode error (status: 500, message: Error: The current identity has not been registered)
Command failed
```

## Steps to Reproduce

1. Start with an already deployed Business Network.
2. Assign an identity to a participant.
3. Assign that identity to a wallet (I used the REST API using _passport-google-oauth_ as the authentication and `POST /wallets/{id}/identities`.)
4. Set the identity to be the default wallet identity. (I used `POST /wallets/{id}/identities/{fk}/setDefault`.)
5. Test out a command, eg. _ping_ `GET /system/ping`.
6. Now go off and do something useful, get a coffee, chat with some co-workers, book a vacation, wait a bit.
8. When you come back test out a command, eg. _ping_ `GET /system/ping`.

When running the command after I've returned to the API explorer after an unknown amount of time this is when I get the error `Error: Error trying to ping. Error: Error trying to query chaincode. Error: chaincode error (status: 500, message: Error: The current identity has not been registered)`.

## Existing issues

- [Stack Overflow issues](http://stackoverflow.com/tags/hyperledger-composer)
- [GitHub Issues](https://github.com/hyperledger/composer/issues)
- ✓ [Rocket Chat history](https://chat.hyperledger.org/channel/composer)

My [initial message](https://chat.hyperledger.org/channel/composer?msg=kRKeWxWhCSQyetnK2) regarding the issue.  [The response](https://chat.hyperledger.org/channel/composer?msg=fpchkwZdWaWon4ie6).

## Context

I was trying to have a participant identity participate in using my business network, reading assets and making transactions.

## Your Environment

Followed the [Composer docs](https://hyperledger.github.io/composer/installing/development-tools.html) to setup my Fabric network.
* Version used
  * Composer v0.10.1
  * Composer REST server v0.10.1
* Environment name and version (e.g. Chrome 39, node.js 5.4)
  * Chrome 59.0.3071.115 (Official Build) (64-bit) (Windows)
  * NodeJS v6.11.1 (Linux)
  * npm v5.3.0 (Linux)
* Operating System and version (desktop or mobile): Ubuntu 16.04.2 LTS

**Comment by ScottMorris** [comment]
Still seeing the issue, it seems like the identity somehow times out after it has been issued.  Issuing a new one works but then after some period of time (a few minutes) it doesn't work anymore.

**Comment by davidkel** [comment]
@ScottMorris Hi, would you be able to post the chaincode container logs of your business network that exhibits the problem ? I haven't had a chance to try to recreate the issue yet, but maybe the logs contain a clue.

**Comment by davidkel** [comment]
I believe this to a be symptom of #1761. What I believe happens is that the wallet is held in memory but after a period of inactivity the connection is terminated and the originally loaded certificates are lost, but the identity is still registered with the original certificate. When you attempt another ping the userid is re-enrolled and that results in a different certificate and now that certificate doesn't compare with the originally registered certificate and that results in the error.

**Comment by davidkel** [comment]
So I am thinking about this more. It would seem that the wallet must still contain the userid and secret, but somehow the certificate data has been lost such that it enrolls again. The Wallet and the entry in the wallet must still be there and it looks like that cerificate info is stored with the wallet identity. Although #1761 would solve the problem in this case, if the rest server remained up then this should not have required a new enroll to occur.

**Comment by ScottMorris** [comment]
@davidkel Sorry for the delay, it was a holiday here in Ontario and I was away all weekend.  I'll try to get the CA and chaincode logs and see if they contain any information.

One thing I've noticed is if the RESTserver is rebooted the wallet information is lost.  It appears to hold it in memory instead of persisting it on disk.  Also, if I use the REST server as _admin_ I can see the issued Identities, but if I try to associate them with a wallet they still fail with the same ``The current identity has not been registered`` error.

**Comment by davidkel** [comment]
@ScottMorris were you using just the in memory capability or were you using mongoDB to persist ? I haven't been able to recreate it with the in memory capability, but another user had exactly the same problem with mongoDB and they tried again using the in memory one and the problem didn't occur.
One other thing, there is an issue with the identity service. So when you perform your testing, once you have issued the identity, don't use the command line after that but use the rest server to ping using that identity and continue just using the rest server. If you try to use the command line it will return a 'not registered' error.

**Comment by davidkel** [comment]
Ok, other user has now reported that the same problem also happens with the in memory implementation as well.

**Comment by ScottMorris** [comment]
@davidkel I didn't setup Mongo so I don't think i'm using it.  As for the testing, I've been using the REST api to do all the work.  I tried to cli but I only tried it to see if the error was only output by the api.  I'm looking into getting the logs right now.

**Comment by ScottMorris** [comment]
I've grabbed the logs from the CA.  Over the weekend my server rebooted and the network wasn't connectable anymore, when I started it by running ./startFabric.sh it whipped out my old docker images.  After this refresh things seem to be stable...I wish I still had a copy of the old docker images since they contained the issue.

However, the log here is from before I accidentally wiped the images, so hopefully it contains some clues.
[logs.txt](https://github.com/hyperledger/composer/files/1209193/logs.txt)

**Comment by ScottMorris** [comment]
It appears to have happened again to my co-worker.  I've tried to get out of him a play by play based on his actions and what appears in the CA Log.  I've added the pertanatnt section of the new log to the bottom of this post, as well as the file.

His steps
1.  [18:38 (log time)] He created a participant and then assigned it the identity _bob_.
    * It seems this is the ``POST /api/v1/register`` call.
2. [18:40] He assigned the identity _bob_ to a wallet (using Passport-Google-Oauth as the auth scheme).  He then set _bob_ as the default.
    * It seems the assignment to the wallet is the ``POST /api/v1/enroll`` call.
3. For a few minutes he made various API requests that all succeeded.
4. [18:50] Around this time his API request failed with the error, ``The current identity has not been registered`` and in the log it shows that another enrolment was performed, which issued a new certificate.
    * The second ``POST /api/v1/enroll``.

The relevant piece of the log file
```
2017/08/08 18:38:45 [DEBUG] Received request
POST /api/v1/register
Authorization: LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUI4RENDQVplZ0F3SUJBZ0lVQjh1YjZVVkpRbCtDSnMzUmoxeEk5OXdJSDBZd0NnWUlLb1pJemowRUF3SXcKY3pFTE1Ba0dBMVVFQmhNQ1ZWTXhFekFSQmdOVkJBZ1RDa05oYkdsbWIzSnVhV0V4RmpBVUJnTlZCQWNURFZOaApiaUJHY21GdVkybHpZMjh4R1RBWEJnTlZCQW9URUc5eVp6RXVaWGhoYlhCc1pTNWpiMjB4SERBYUJnTlZCQU1UCkUyTmhMbTl5WnpFdVpYaGhiWEJzWlM1amIyMHdIaGNOTVRjd09EQTRNVGd3T0RBd1doY05NVGd3T0RBNE1UZ3cKT0RBd1dqQVFNUTR3REFZRFZRUURFd1ZoWkcxcGJqQlpNQk1HQnlxR1NNNDlBZ0VHQ0NxR1NNNDlBd0VIQTBJQQpCS2JLdnlNazNwQmEra09pck94VkhZNURudTQ1SE4yTER5Sm1CSzlWTUI0Y3EwMGVyZ3d4OUcrTTJiRWo3c1BjCis2L0FDanFOS1lxUDFOY0wvMkRBK1QramJEQnFNQTRHQTFVZER3RUIvd1FFQXdJSGdEQU1CZ05WSFJNQkFmOEUKQWpBQU1CMEdBMVVkRGdRV0JCUVdFS3J2dmgvVlQ1QXhFTXZxdWNHbTQwV2cvakFyQmdOVkhTTUVKREFpZ0NBWgpxMldydXdTQWZhMFM1TUNwcXFaa25uQ0dqanE5QWhlakl0aWVSK0dtcmpBS0JnZ3Foa2pPUFFRREFnTkhBREJFCkFpQTBtN2FReUVZKzRNTER6TUxFVEIvM1VBYzFmOFRTdHp1V3huSnFQdHBET1FJZ085V1FDVG5IRG1RZm9ZUkoKMU54S3lFcktvcTJDWDdsN3BOaVo3dzRlSFhVPQotLS0tLUVORCBDRVJUSUZJQ0FURS0tLS0tCg==.MEQCIFWOBcyHf29htvXDh9KB/p9U+mBuOPZMoylmB1yEMtJjAiAiXBgJI+qCJQksEaF38G/7LXbMtzsHXKOYWnXkXbwd0A==
{"id":"bob","type":"client","affiliation":"org1","max_enrollments":0,"attrs":[],"caName":"ca.org1.example.com"}
2017/08/08 18:38:45 [DEBUG] Directing traffic to CA ca.org1.example.com
2017/08/08 18:38:45 [DEBUG] Checking for revocation/expiration of certificate owned by 'admin'
2017/08/08 18:38:45 [DEBUG] DB: Get certificate by serial (7cb9be94549425f8226cdd18f5c48f7dc081f46) and aki (19ab65abbb04807dad12e4c0a9aaa6649e70868e3abd0217a322d89e47e1a6ae)
2017/08/08 18:38:45 [DEBUG] Successful authentication of 'admin'
2017/08/08 18:38:45 [DEBUG] Register request received
2017/08/08 18:38:45 [DEBUG] Received registration request from admin: &{RegistrationRequest:{Name:bob Type:client Secret:<<user-specified>> MaxEnrollments:0 Affiliation:org1 Attributes:[] CAName:ca.org1.example.com}}
2017/08/08 18:38:45 [DEBUG] canRegister - Check to see if user admin can register
2017/08/08 18:38:45 [DEBUG] DB: Getting identity admin
2017/08/08 18:38:45 [DEBUG] Validate ID
2017/08/08 18:38:45 [DEBUG] An affiliation is required for identity type client
2017/08/08 18:38:45 [DEBUG] Validating affiliation: org1
2017/08/08 18:38:45 [DEBUG] DB: Get affiliation org1
2017/08/08 18:38:45 [DEBUG] Registering user id: bob
2017/08/08 18:38:45 [DEBUG] Max enrollment value verification - User specified max enrollment: 0, CA max enrollment: -1
2017/08/08 18:38:45 [DEBUG] DB: Getting identity bob
2017/08/08 18:38:45 [DEBUG] DB: Add identity bob
2017/08/08 18:38:45 [DEBUG] Successfully added identity bob to the database
2017/08/08 18:38:45 [DEBUG] Registration completed - sending response 0xc420024260
2017/08/08 18:38:45 [INFO] 172.18.0.1:52754 - "POST /api/v1/register" 200
2017/08/08 18:40:04 [DEBUG] Received request
POST /api/v1/enroll
Authorization: Basic Ym9iOlF2dEdYQ3JLbE9peg==
{"caName":"ca.org1.example.com","certificate_request":"-----BEGIN CERTIFICATE REQUEST-----\r\nMIHKMHACAQAwDjEMMAoGA1UEAwwDYm9iMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcD\r\nQgAEeUgxhtOHMRVCVm1KJjbnv7Ni21eE4PWLOsNe66STr7frrdAarkqMCOPwfwzm\r\nEVQSm/PA3Tlj9K88m5hKKTrHeKAAMAwGCCqGSM49BAMCBQADSAAwRQIhANLEi1+X\r\nLkH5sxLme2/xaSjUXCPmnU/M8PGi5I3MMcZnAiB03EyLbl9HjqYzQ57kXLKN5o9Y\r\nM5bRGAJy6yMyWM5zVQ==\r\n-----END CERTIFICATE REQUEST-----\r\n"}
2017/08/08 18:40:04 [DEBUG] Directing traffic to CA ca.org1.example.com
2017/08/08 18:40:04 [DEBUG] DB: Getting identity bob
2017/08/08 18:40:04 [DEBUG] DB: Login user bob with max enrollments of -1 and state of 0
2017/08/08 18:40:04 [DEBUG] Successfully incremented state for identity bob to 1
2017/08/08 18:40:04 [DEBUG] DB: identity bob successfully logged in
2017/08/08 18:40:04 [DEBUG] Identity/Pass was correct
2017/08/08 18:40:04 [DEBUG] Received request for endpoint enroll
2017/08/08 18:40:04 [DEBUG] Enrollment request: {SignRequest:{Hosts:[] Request:-----BEGIN CERTIFICATE REQUEST-----
MIHKMHACAQAwDjEMMAoGA1UEAwwDYm9iMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcD
QgAEeUgxhtOHMRVCVm1KJjbnv7Ni21eE4PWLOsNe66STr7frrdAarkqMCOPwfwzm
EVQSm/PA3Tlj9K88m5hKKTrHeKAAMAwGCCqGSM49BAMCBQADSAAwRQIhANLEi1+X
LkH5sxLme2/xaSjUXCPmnU/M8PGi5I3MMcZnAiB03EyLbl9HjqYzQ57kXLKN5o9Y
M5bRGAJy6yMyWM5zVQ==
-----END CERTIFICATE REQUEST-----
 Subject:<nil> Profile: CRLOverride: Label: Serial:<nil> Extensions:[]} CAName:ca.org1.example.com}
2017/08/08 18:40:04 [DEBUG] csrAuthCheck: enrollment ID=bob, CommonName=bob, Subject=<nil>
2017/08/08 18:40:04 [DEBUG] CSR request received
2017/08/08 18:40:04 [INFO] signed certificate with serial number 727793618108623714957241197725696348146281975886
2017/08/08 18:40:04 [DEBUG] DB: Insert Certificate
2017/08/08 18:40:04 [DEBUG] Saved serial number as hex 7f7b6500a0c4af19748a770d13de0cd0d377904e
2017/08/08 18:40:04 [DEBUG] saved certificate with serial number 727793618108623714957241197725696348146281975886
2017/08/08 18:40:04 [INFO] 172.18.0.1:52758 - "POST /api/v1/enroll" 200
2017/08/08 18:50:56 [DEBUG] Received request
POST /api/v1/enroll
Authorization: Basic Ym9iOlF2dEdYQ3JLbE9peg==
{"caName":"ca.org1.example.com","certificate_request":"-----BEGIN CERTIFICATE REQUEST-----\r\nMIHKMHACAQAwDjEMMAoGA1UEAwwDYm9iMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcD\r\nQgAE5KA6nwFKI0J+BYf/Kqi0s5AuZWIwJ0jlDosVb/vs3eBaSrRz8Ri3TZqZHpE+\r\nMjRxrw2Nmeao7Dk4G8v/6d1uvaAAMAwGCCqGSM49BAMCBQADSAAwRQIhANR5Xudt\r\nw0ptEjRcJQ6S2Bbnt7NTRbLDufm+U3E+lumlAiAnO/e8GQPluzUWrxEAqf2APZv1\r\n9dtlxIRxkysAm9LZWg==\r\n-----END CERTIFICATE REQUEST-----\r\n"}
2017/08/08 18:50:56 [DEBUG] Directing traffic to CA ca.org1.example.com
2017/08/08 18:50:56 [DEBUG] DB: Getting identity bob
2017/08/08 18:50:56 [DEBUG] DB: Login user bob with max enrollments of -1 and state of 1
2017/08/08 18:50:57 [DEBUG] Successfully incremented state for identity bob to 2
2017/08/08 18:50:57 [DEBUG] DB: identity bob successfully logged in
2017/08/08 18:50:57 [DEBUG] Identity/Pass was correct
2017/08/08 18:50:57 [DEBUG] Received request for endpoint enroll
2017/08/08 18:50:57 [DEBUG] Enrollment request: {SignRequest:{Hosts:[] Request:-----BEGIN CERTIFICATE REQUEST-----
MIHKMHACAQAwDjEMMAoGA1UEAwwDYm9iMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcD
QgAE5KA6nwFKI0J+BYf/Kqi0s5AuZWIwJ0jlDosVb/vs3eBaSrRz8Ri3TZqZHpE+
MjRxrw2Nmeao7Dk4G8v/6d1uvaAAMAwGCCqGSM49BAMCBQADSAAwRQIhANR5Xudt
w0ptEjRcJQ6S2Bbnt7NTRbLDufm+U3E+lumlAiAnO/e8GQPluzUWrxEAqf2APZv1
9dtlxIRxkysAm9LZWg==
-----END CERTIFICATE REQUEST-----
 Subject:<nil> Profile: CRLOverride: Label: Serial:<nil> Extensions:[]} CAName:ca.org1.example.com}
2017/08/08 18:50:57 [DEBUG] csrAuthCheck: enrollment ID=bob, CommonName=bob, Subject=<nil>
2017/08/08 18:50:57 [DEBUG] CSR request received
2017/08/08 18:50:57 [INFO] signed certificate with serial number 395828804141164546460638482149297734289809423186
2017/08/08 18:50:57 [DEBUG] DB: Insert Certificate
2017/08/08 18:50:57 [DEBUG] Saved serial number as hex 455593d3e6d5e9283163d4b2f504c45c89ff3b52
2017/08/08 18:50:57 [DEBUG] saved certificate with serial number 395828804141164546460638482149297734289809423186
2017/08/08 18:50:57 [INFO] 172.18.0.1:52782 - "POST /api/v1/enroll" 200
```
The full log file, [logs-2017-08-08-14-52-GMT-4.txt](https://github.com/hyperledger/composer/files/1209257/logs-2017-08-08-14-52-GMT-4.txt)

**Comment by sstone1** [comment]
@ScottMorris should be fixed in v0.11.3, please can you give it a try?
